// image-gallery/components/Button.js
"use strict";
import PropTypes from 'prop-types';
import React from 'react';
import { TouchableOpacity } from 'react-native';
import FAIcon from 'react-native-vector-icons/FontAwesome';

class Button extends React.Component {

	static displayName = "Button";

	static propTypes = {
		style: PropTypes.object,
		onPress: PropTypes.func,
		iconName: PropTypes.string.isRequired,
		iconStyle: PropTypes.object,
		hitSlop: PropTypes.object
	};

	static defaultProps = {
		hitSlop: {
			top: 10,
			right: 10,
			bottom: 10,
			left: 10
		}
	};

	shouldComponentUpdate( nextProps ) {
		
		return (
			this.props.iconName !== nextProps.iconName ||
			this.props.onPress != nextProps.onPress
		);
	}

	render() {

		const { iconName, style, iconStyle, hitSlop, onPress } = this.props;

		return (
			<TouchableOpacity onPress={ onPress } style={[ _styles.container, style ]} hitSlop={ hitSlop }>
				<FAIcon name={ iconName } style={ [_styles.icon, iconStyle] }/>
			</TouchableOpacity>
		)
	}
}

const _styles = {
	container: {
		width: 30,
		height: 30,
		borderWidth: 0.5,
		borderColor: "white",
		borderRadius: 15,
		justifyContent: "center",
		alignItems: "center"
	},
	icon: {
		fontSize: 16,
		color: "white"
	}
};

export default Button;
