// image-gallery/components/Navigation.js
"use strict";
import PropTypes from 'prop-types';
import React from 'react';
import { View } from 'react-native';
import Button from './Button';

class Navigation extends React.Component {

	static displayName = "Navigation";

	static propTypes = {
		style: PropTypes.object,
		buttonStyle: PropTypes.object,
		iconStyle: PropTypes.object,
		onPrev: PropTypes.func.isRequired,
		onNext: PropTypes.func.isRequired,
		isPlay: PropTypes.bool,
		onPause: PropTypes.func.isRequired,
		onPlay: PropTypes.func.isRequired,
		hitSlop: PropTypes.object
	};

	shouldComponentUpdate( nextProps ) {
		
		return (
			this.props.isPlay !== nextProps.isPlay
		);
	}

	render() {

		const { 
			style, 
			buttonStyle, 
			iconStyle,
			isPlay,
			onPrev,
			onNext,
			onPause,
			onPlay,
			hitSlop
		} = this.props;

		return (
			<View style={[ _styles.container, style ]} pointerEvents="box-none">
				<Button 
					iconName    = "angle-left" 
					style       = { buttonStyle } 
					iconStyle   = { iconStyle }
					onPress     = { onPrev }
					hitSlop 	= { hitSlop }
				/>
				<Button 
					iconName 	= { isPlay ? "pause" : "play" }
					style 		= { buttonStyle } 
					iconStyle 	= { iconStyle }
					onPress 	= { isPlay ? onPause : onPlay }
					hitSlop 	= { hitSlop }
				/>
				<Button 
					iconName 	= "angle-right" 
					style 		= { buttonStyle } 
					iconStyle 	= { iconStyle }
					onPress 	= { onNext }
					hitSlop 	= { hitSlop }
				/>
			</View>
		)
	}
}

const _styles = {
	container: {
		position: "absolute",
		flexDirection: "row",
		bottom: 0,
		left: 0,
		right: 0,
		padding: 10,
		backgroundColor: "rgba(0, 0, 0, 0.5)",
		alignItems: "center",
		justifyContent: "space-around"
	}
};

export default Navigation;
